lighthouse-core/audits/byte-efficiency/uses-responsive-images-snapshot.js

```diff
@@ -0,0 +1,93 @@
+/**
+ * @license Copyright 2021 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+/**
+ * @fileoverview Checks to see if the images used on the page are larger than
+ *   their display sizes. The audit will list all images that are larger than
+ *   their display size with DPR (a 1000px wide image displayed as a
+ *   500px high-res image on a Retina display is 100% used);
+ */
+'use strict';
+
+const Audit = require('../audit.js');
+const UsesResponsiveImages = require('./uses-responsive-images.js');
+const URL = require('../../lib/url-shim.js');
+const i18n = require('../../lib/i18n/i18n.js');
+
+const UIStrings = {
+  /** Label for a column in a data table; entries will be the dimensions of an image as it appears on the page. */
+  columnDisplayedDimensions: 'Displayed dimensions',
+  /** Label for a column in a data table; entries will be the dimensions of an image from it's source file. */
+  columnActualDimensions: 'Actual dimensions',
+};
+
+Object.assign(UIStrings, UsesResponsiveImages.UIStrings);
```

Comment:
Is there a decent way to share UI strings between files without putting them in `i18n.UIStrings`? This is pretty sketchy, I'm not certain the imported strings will even be translated.

Reply:
They would be translated just the same. The collect-strings script requires each module, so L26 is actually executed as you'd expect.

This is novel, but seems good to me.

Reply:
No, we are only using `title` and `description` which are specific to these two files.

Reply:
They'll be translated, but because they're in `UIStrings` in `uses-responsive-images.js`, not in here. `collect-strings` does require the module, but only for [getting each string _value_](https://github.com/GoogleChrome/lighthouse/blob/d8d1d39f96edb1df41855fc29f3b0ae91c76096b/lighthouse-core/scripts/i18n/collect-strings.js#L503-L504), not for getting the list of strings to include (that comes from tsc parsing of the regex-exptracted local `UIStrings` object literal).

However, `title` and `description` still won't ever be localized because `__filename` in this file results in ids like `.../uses-responsive-images-snapshot.js | title` and won't match `.../uses-responsive-images.js | title`, so it'll always fall back to the english string added in the `Object.assign()` call.

There's not really a pretty way of doing this. We wanted strings to be defined in the file where they are used and we wanted to not have to come up with globally-unique names, so this is one of the constraints.

I'm not sure if we want to go down this path and it may be opening pandora's box just by mentioning this, but....there's no reason why the first argument of `i18n.createMessageInstanceIdFn()` _has_ to be `__filename`. It could also be `const str_ = require.resolve('./uses-responsive-images.js', UsesResponsiveImages.UIStrings)` and all the strings could be over in the original file. Or you could have two `str_`s, one for local strings and one for the other file's.

Please consider that these might be terrible ideas before trying them, though :) It doesn't seem obviously dangerous, but it does e.g. start to introduce shadow file dependencies that are much harder to analyze and refactor than the normal require/import ones (and we don't currently have any test coverage for making sure all audits are translated, just that they're translatable as long as they're following the same pattern as the rest). If it gets complicated it may be a major pain to unwind.

Reply:
I've changed the implementation to what @brendankenny is talking about. The UI strings are translated as we expect. I also tried to highlight the new dependency by importing the `str_` function from `UsesResponsiveImages`.

Reply:
Nice! It still feels a little weird, but `str_` is designed to provide portable results (within reason) and this is a nice way to keep the dependencies explicit.